### PR TITLE
SAAS-3847: Added AWS Backup Compliant Lifecycle Configured Plugin

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

File: plugins/aws/backup/compliantLifecyleConfigured.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                        region, resource);
+                } 
+
+                if (getBackupPlan.data.BackupPlan &&
+                getBackupPlan.data.BackupPlan.Rules[0] &&
+                getBackupPlan.data.BackupPlan.Rules[0].Lifecycle &&
+                getBackupPlan.data.BackupPlan.Rules[0].Lifecycle.DeleteAfterDays == null &&
+                getBackupPlan.data.BackupPlan.Rules[0].Lifecycle.MoveToColdStorageAfterDays == null) {
+                    helpers.addResult(results, 2,
+                        'there is no lifecycle configuration enabled for the selected Amazon Backup plan', region, resource);
```
Comment (sadeed12345) [inline plugins/aws/backup/compliantLifecyleConfigured.js:56]:
```suggestion
                        'No lifecycle configuration enabled for the selected Amazon Backup plan', region, resource);
```

File: plugins/aws/backup/compliantLifecyleConfigured.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Backup Vault Encrypted',
+    category: 'Backup',
+    domain: 'Storage',
+    description: 'Ensure that a compliant lifecycle configuration is enabled for your Amazon Backup plans in order to meet compliance requirements when it comes to security and cost optimization.',
```
Comment (sadeed12345) [inline plugins/aws/backup/compliantLifecyleConfigured.js:8]:
```suggestion
    description: 'Ensure that a compliant lifecycle configuration is enabled for your Amazon Backup plans in order to meet compliance requirements when it comes to security and cost optimization. ',
```

File: plugins/aws/backup/compliantLifecyleConfigured.spec.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(3);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('Unable to query Backup plans')
+                done();
+            });
+        });
+
+        it('should UNKNOWN if Unable to get Backup plan description', function (done) {
+            const cache = createCache([listBackupPlans[0]], null, { message: "Unable to get Backup plan description" });
```
Comment (sadeed12345) [inline plugins/aws/backup/compliantLifecyleConfigured.spec.js:133]:
```suggestion
            const cache = createCache([listBackupPlans[0]], null, null, { message: "Unable to get Backup plan description" });
```

File: plugins/aws/backup/compliantLifecyleConfigured.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                getBackupPlan.data.BackupPlan.Rules[0].Lifecycle &&
+                getBackupPlan.data.BackupPlan.Rules[0].Lifecycle.DeleteAfterDays == null &&
+                getBackupPlan.data.BackupPlan.Rules[0].Lifecycle.MoveToColdStorageAfterDays == null) {
+                    helpers.addResult(results, 2,
+                        'No lifecycle configuration enabled for the selected Amazon Backup plan', region, resource);
+                } else {
+                    helpers.addResult(results, 0,
+                        'Lifecycle configuration enabled for the selected Amazon Backup plan', region, resource);
+                }
+
```
Comment (sadeed12345) [inline plugins/aws/backup/compliantLifecyleConfigured.js:61]:
```suggestion
```

File: plugins/aws/backup/compliantLifecyleConfigured.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        async.each(regions.backup, function(region, rcb) {
+            var listBackupPlans = helpers.addSource(cache, source,
+                ['backup', 'listBackupPlans', region]);
+
+            if (!listBackupPlans) return rcb();
+
+            if (listBackupPlans.err || !listBackupPlans.data) {
+                helpers.addResult(results, 3,
+                    'Unable to query Backup plans: ' + helpers.addError(listBackupPlans), region);
```
Comment (sadeed12345) [inline plugins/aws/backup/compliantLifecyleConfigured.js:28]:
```suggestion
                    'Unable to list Backup plans: ' + helpers.addError(listBackupPlans), region);
```

File: plugins/aws/backup/compliantLifecyleConfigured.spec.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+            compliantLifecyleConfigured.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('No Backup plans found')
+                done();
+            });
+        });
+
+        it('should UNKNOWN if Unable to query Backup plans', function (done) {
```
Comment (sadeed12345) [inline plugins/aws/backup/compliantLifecyleConfigured.spec.js:121]:
```suggestion
        it('should UNKNOWN if Unable to list Backup plans', function (done) {
```

File: plugins/aws/backup/compliantLifecyleConfigured.spec.js [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('No Backup plans found')
+                done();
+            });
+        });
+
+        it('should UNKNOWN if Unable to list Backup plans', function (done) {
+            const cache = createCache(null, null, { message: "Unable to query Backup plans" });
```
Comment (sadeed12345) [inline plugins/aws/backup/compliantLifecyleConfigured.spec.js:122]:
```suggestion
            const cache = createCache(null, null, { message: "Unable to list Backup plans" });
```

File: plugins/aws/backup/compliantLifecyleConfigured.spec.js [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+            });
+        });
+
+        it('should UNKNOWN if Unable to list Backup plans', function (done) {
+            const cache = createCache(null, null, { message: "Unable to list Backup plans" });
+            compliantLifecyleConfigured.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(3);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('Unable to query Backup plans')
```
Comment (sadeed12345) [inline plugins/aws/backup/compliantLifecyleConfigured.spec.js:127]:
```suggestion
                expect(results[0].message).to.include('Unable to list Backup plans')
```

File: plugins/aws/backup/compliantLifecyleConfigured.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Backup Vault Encrypted',
```
Comment (sadeed12345) [inline plugins/aws/backup/compliantLifecyleConfigured.js:5]:
```suggestion
    title: 'AWS Backup Compliant Lifecycle Configured',
```

File: plugins/aws/backup/compliantLifecyleConfigured.js [hunk 10]
```diff
@@ -0,0 +1,68 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'AWS Backup Compliant Lifecycle Configured',
+    category: 'Backup',
+    domain: 'Storage',
+    description: 'Ensure that a compliant lifecycle configuration is enabled for your Amazon Backup plans in order to meet compliance requirements when it comes to security and cost optimization. ',
+    more_info: 'The AWS Backup lifecycle configuration contains an array of Transition objects specifying how long in days before a recovery point transitions to cold storage or is deleted.'+ 
+        'DeleteAfterDays specifies the number of days after creation that a recovery point is deleted. MoveToColdStorageAfterDays Specifies the number of days after creation that a recovery point is moved to cold storage.',
```
Comment (AkhtarAmir) [inline plugins/aws/backup/compliantLifecyleConfigured.js:10]:
```suggestion

```

File: plugins/aws/backup/compliantLifecyleConfigured.js [hunk 11]
```diff
@@ -0,0 +1,68 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'AWS Backup Compliant Lifecycle Configured',
+    category: 'Backup',
+    domain: 'Storage',
+    description: 'Ensure that a compliant lifecycle configuration is enabled for your Amazon Backup plans in order to meet compliance requirements when it comes to security and cost optimization. ',
+    more_info: 'The AWS Backup lifecycle configuration contains an array of Transition objects specifying how long in days before a recovery point transitions to cold storage or is deleted.'+
```
Comment (AkhtarAmir) [inline plugins/aws/backup/compliantLifecyleConfigured.js:9]:
```suggestion
    more_info: 'The AWS Backup lifecycle configuration contains an array of transition objects specifying how long in days before a recovery point transitions to cold storage or is deleted.'+ 
```

File: plugins/aws/backup/compliantLifecyleConfigured.js [hunk 12]
```diff
@@ -0,0 +1,68 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'AWS Backup Compliant Lifecycle Configured',
+    category: 'Backup',
+    domain: 'Storage',
+    description: 'Ensure that a compliant lifecycle configuration is enabled for your Amazon Backup plans in order to meet compliance requirements when it comes to security and cost optimization. ',
```
Comment (AkhtarAmir) [inline plugins/aws/backup/compliantLifecyleConfigured.js:8]:
```suggestion
    description: 'Ensure that a compliant lifecycle configuration is enabled for your Amazon Backup plans in order to meet compliance requirements when it comes to security and cost optimization.',
```

File: plugins/aws/backup/compliantLifecyleConfigured.js [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    more_info: 'The AWS Backup lifecycle configuration contains an array of Transition objects specifying how long in days before a recovery point transitions to cold storage or is deleted.'+ 
+        'DeleteAfterDays specifies the number of days after creation that a recovery point is deleted. MoveToColdStorageAfterDays Specifies the number of days after creation that a recovery point is moved to cold storage.',
+    recommended_action: 'Enable compliant lifecycle configuration for your Amazon Backup plans in order to meet compliance requirements',
```
Comment (AkhtarAmir) [inline plugins/aws/backup/compliantLifecyleConfigured.js:11]:
```suggestion
    recommended_action: 'Enable compliant lifecycle configuration for your Amazon Backup plans',
```

File: plugins/aws/backup/compliantLifecyleConfigured.js [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+                if (!getBackupPlan || getBackupPlan.err || !getBackupPlan.data) {
+                    helpers.addResult(results, 3,
+                        `Unable to get Backup plan description: ${helpers.addError(getBackupPlan)}`,
+                        region, resource);
+                } 
+
+                if (getBackupPlan.data.BackupPlan &&
+                getBackupPlan.data.BackupPlan.Rules[0] &&
+                getBackupPlan.data.BackupPlan.Rules[0].Lifecycle &&
```
Comment (AkhtarAmir) [inline plugins/aws/backup/compliantLifecyleConfigured.js:51]: What is the length of getBackupPlan.data.BackupPlan.Rules is more than 1?

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

**Review by AkhtarAmir — APPROVED** [review]

**Pull request merged**